hypothesis-python/src/hypothesis/strategies/_internal/numbers.py

```diff
@@ -416,10 +439,37 @@ def floats(
     elif max_value == -math.inf:
         raise InvalidArgument("allow_infinity=False excludes max_value=-inf")
 
+    smallest_normal = width_smallest_normals[width]
+    if allow_subnormal is None:
+        if min_value is not None and max_value is not None:
+            if min_value == max_value:
+                allow_subnormal = -smallest_normal < min_value < smallest_normal
+            else:
+                allow_subnormal = (
+                    min_value < smallest_normal and max_value > -smallest_normal
+                )
+        elif min_value is not None:
+            allow_subnormal = min_value < smallest_normal
+        elif max_value is not None:
+            allow_subnormal = max_value > -smallest_normal
+        else:
+            allow_subnormal = True
+    if allow_subnormal and min_value is not None and min_value >= smallest_normal:
+        raise InvalidArgument(
+            f"allow_subnormal=True, but minimum value {min_value} "
+            f"excludes values below smallest positive normal {smallest_normal}"
+        )
+    if allow_subnormal and max_value is not None and max_value <= -smallest_normal:
+        raise InvalidArgument(
+            f"allow_subnormal=True, but maximum value {max_value} "
+            f"excludes values above smallest negative normal {-smallest_normal}"
+        )
```

Comment:
These messages need to mention the float width, since the value of smallest_normal varies.